### ISSUE TITLE
Implement review rating test and learn

### ISSUE THREAD

**Issue by hdrury1** [sciety/sciety#303]
## Scenarios

- [x] Given I haven't voted before
  - [x] When I express an upvote,
    - [x] Then the upvote counter increments 
    - [x] and the upvote button changes
  - [x] When I express an downvote,
    - [x]  Then the downvote counter increments 
    - [x] and the downvote button changes
- [x] Given I have upvoted before
  - [x] When I express an upvote on an outdated UI
     - [x] Then the upvote counter stays the same and the upvote button doesn't change
  - [x] When I express a downvote on an outdated UI
    - [x] Then the downvote counter stays the same and the downvote button doesn't change
- [x] Given I have upvoted before, When I click upvote, Then the upvote counter decrements and the upvote button changes
- [x] Given I have downvoted before, When I click upvote, Then the downvote counter decrements and the upvote counter increments and the buttons change
- [x] Given I have downvoted before, When I click downvote, Then the downvote counter decrements and the downvote button changes
- [x] Given I am not logged in
  - [x] given I haven't responded before when I respond `is helpful`
    - [x] redirect to login page
    - [x] returned to article page I was on when I responded
    - [x] the response is recorded
  - [x] on article, respond `helpful` to review one, cancel log in, visit page again, respond `helpful` to review two, complete the login
    - [x] returned to article page I was on when I responded
    - [x] only record response to review two
  - [x] on page, already responded `helpful`, log out in other tab, invoke `revoke` on page where UI indicates I am still logged in (outdated)
    - [x] redirect to login page
    - [x] revoke is recorded

### Middlewares ToDo

- [x] save the command on `/respond`
- [x] replay the command on twitter callback

### Testing

- options
  - ~unit tests: single middleware~: poor coverage given the complex flow involving redirects and authentication
  - [x] ~request-scoped integration tests: put something in the session, check redirect~: rabbit hole in stubbing Twitter
  - [x] manual browser testing for scenarios by human clicking on things
  - ~non-JavaScript executing headless scraper (a la Goutte)~
  - ~browser-based functional tests: click a thing, log in, come back, check the result~
  - ~cross-browser tests: do one thing on desktop Chrome, mobile Firefox, and more; supporting local use of emulated browsers~

## Bugs

- [x] interaction on different review in `project-user-review-response`
- [x] `project-user-review-response` doesn't cater for revoked not helpful events
- [x] check the other projection too

## Language

- [x] language of user journey (UX) vs language of `Voting` Bounded Context

## Pre-release

- [x] new Domain Events should be persisted when stable rather than kept in memory and lost on restart
  - [x] deserialization code
  - [x] ~look into removing the duplication in both event types~ put some TypeScript-side refactoring into post-release
- [x] ensure tag manager trigger has been added
  - [x] track clicks on respond-helpful/revoke, respond-not-helpful/revoke
  - [x] Trigger: ~Click All Elements / Some Clicks on our custom variable, when it matches `.+`~ `Form Submit` when `Form URL` contains `/respond`
  - [x] Variable: ~Auto-Event Variable, Element Attribute is `aria-label`~ Data Layer Variable capturing the `value` of the second element of the form
  - [x] Tag 
    - [x] single category `Breadcrumb`
    - [x] actions `respond-helpful` or `respond-not-helpful`, should come from the Variable
    - [x] no label
  - [x] what is the tag `GA4 - Pageview`? Removed by @hdrury1 
- what to do with login?
  - ~release as-is: login sends you back to the page without memory of what you clicked~
  - ~option 1, the original scope: "replay command after login" scenarios, with automated browser tests~
  - [x] option 1-, the original scope: "replay command after login" scenarios, manual testing only
  - ~option 2, the lot of extra work: introduce a generated identity and use it in place of log in~
  - ~option 3: only show buttons to logged in users, and do nothing else~
  - ~option 3+: show buttons to logged in users and "log in to see our sciency buttons!" call to action to everyone else~
- [x] remove from feature flag

## Post-release

- [x] Taiko-based automated browser tests
  - [x] ad-hoc-executed browser tests with `make dev` and Taiko, but not in CI
  - [x] login scenario (and we decided to keep it)
  - [x] first respond while authenticating scenario
  - Improvements aka make it pretty
    - [x] `make regression` target to run `taiko` and `backstop`
    - [x] use Jest
    - [x] convert to TypeScript
    - [x] apply the same linting rules
      - [x] mount ~need to copy~ `feature-test/` to the container
      - [x] ~eslint~ tsc picks up Taiko in `node_modules` installing `devtools-protocol` as a dev dependencies to get its types, following up with https://github.com/getgauge/taiko/issues/1751
      - [x] use `external-types` instead of installing the `devtools-protocol` package (reduce the overhead of installing a module e.g. mismatched version)
      - [x] why are there 2 separate lint targets for `prod` and `dev`?
    - ~run inside a container~
    - [x] `sleep` is sometimes too short: wait on the application to be ready before starting Jest
    - [x] Don't install chromium in dev container
- [x] replay-command middlewares
  - [x] remove `create` from the middleware of the respond area, and make them named exports
  - [x] to check: various middlewares write to the `command` field --is this okay?
  - [x] duplication between `respond/index` and `respond/finish-respond-command`
  - [x] check for undefined values from the session during twitter callback
  - [x] validate session content together with command
- ~error cases: what to display if votes can't be retrieved? They come from a database, or in-memory after reading it on bootstrap~
- [x] consider raising the TypeScript types abstraction to a single UserRespondedEvent with `"type": 'UserFoundReviewHelpful' | 'UserFoundReviewNotHelpful' | ...`, without affecting persistence
- [x] Return to displaying the review that was voted on (add page anchor)
  - [x] add an anchor to reviews
  - ~escape anchors that contain characters such as `&` that are not safe in an HTML attribute value (non-exhaustive example)~ deferred as being unlikely
  - [x] pass in `APP_SECRET` in staging and prod
  - [x] redirect to anchor within `/respond` (when you are logged in)
    - [x] refactor test not to hardcode session values and signature
    - [x] handling lack of `Referer`: default to homepage
    - ~refactor to use `targetFragmentId`?~
  - [x] redirect to anchor from replay middleware (when you are not logged in)
    - [x] don't put `targetFragmentId` in the session if it doesn't need to survive the request scope
- [x] Styling
- [x] Add helpful button and unhelpful button
  - [x] white space
  - [x] take decision on primary colors
  - [x] keep at the bottom of the review (outside collapsible section) for all screen widths
  - [x] change "Did you find this review helpful?" to "Did you find this evaluation helpful?"
  - [x] what should happen after the page refreshes?
    - [x] ~keep expanded reviews open on  post-response page refresh~ it's out of scope
- [x] Check teaser text length is same as it was
- [x] make `Doi.toString()` add the `doi:` prefix for consistency with `HypothesisAnnotationId` and to make the `/upvote` form `hidden` input work
- [x] consider if feature flag should be a runtime parameter? It shouldn't because we will just rollout to everyone
- [x] when rolling out, don't filter out `.responses` in `backstop.json`
- [x] Refactoring SASS
  - [x] extract a partial for the `article-page` SASS
  - [x] tidy up grid SASS
  - [x] not understanding the two nested grid on the article page i.e. could we remove `div.main-content--article` and use only the grid on `article`?
  - [x] rename `article-feed-section` to `article-feed`, and rename all of its children according to BEM principles
- [x] General refactoring
  - [x] extract a function that goes from `allEvents` to the state of a single review response.
    - [x] test it
  - [x] commands shouldn't know about `GetAllEvents`
  - [x] do something better than filtering over all events multiple times.
  - [x] refactor command to use a `switch` rather than multiple `ifs`
  - [x] Try removing types in signatures not needed by the compiler
  - [x] remove duplication (or unnecessary mentions) of `review-response` triple union type
  - [x] `switch` on command should not have a default
  - [x] command handlers don't need to be curried
  - [x] de-duplicate creation of events in `src` similarly to `makeEvent` in tests
    - [x] use the new factories in the tests
  - [x] (controversial) experiment with replacing the switch with a JavaScript object lookup or some `fp-ts` idiom
  - [x] `// TODO: this should only produce revoke events`
  - [x] coupling of ReviewId and UserId across `review-response` and the commands: the same values must be passed in in 4 places
  - [x] Make function signatures symmetrical for commands
  - [x] ~Command signatures could return a looser or stricter set of events?~
  - [x] Remove default exports in the `respond` folder
- [x] `render-responses`
  - [x] extract and inject
  - [x] make `async`
  - [x] add unit test
  - [x] add port
  - [x] extend port with an identifier
  - [x] Test cases
     - [x] no votes
     - [x] user has upvoted
     - [x] user has downvoted   
  - [x] should pass a UserId to a port, but needs to ask for it to parent components
    - [x] take out `render-response` from `render-review-feed-item` and use from `render-feed`? `reponseHtml` is only appended to the end (before wrapping in `<li>` happens in `render-feed`)
  - [x] propagate `Maybe<UserId>` until here
- [x] `project-votes`
  - [x] unit tests
- [x] adapter for `GetCurrentUserResponses` i.e. `project-current-user-responses`?
- [x] `render-review-feed-item`
  - [x] consider removing duplicate markup
- [x] remove duplication of reviews in render-feed-item.test and handle-article-version-errors.test
- [x] `render-flavour-a-feed` hardcodes the whole feed HTML, so it won't get these. Leave, update or remove?

**Comment by hdrury1** [comment]
Feedback from Naomi indicated that helpful/unhelpful were useful concepts- no immediate concerns.

Feedback from Nick suggested thumbs up/thumbs down were universally accepted patterns for this kind of interface, and should be paired with text for clarity.

**Comment by davidcmoulton** [comment]
Thumb up / down icons sourced from http://materialdesignicons.com/

**Comment by will-byrne** [comment]
Current solution does not trigger off of the button click event and passes through the svg as the element

**Comment by giorgiosironi** [comment]
Found a solution in using a Form Submit trigger along with a brittle variable collecting the `value` of the second input of the form.

**Comment by giorgiosironi** [comment]
**staging** data:
![Screenshot_2020-12-16 Analytics](https://user-images.githubusercontent.com/160299/102370378-e6932c00-3fb4-11eb-8422-84d0bef6b93b.png)

**Comment by hdrury1** [comment]
`GA4 - Pageview` was something I implemented as part of a series of tutorial tasks. Happy for it to be deleted.

**Comment by giorgiosironi** [comment]
Done. Also cleaned up some names of Variables.

**Comment by giorgiosironi** [comment]
Can the other properties under Sciety *(`GA 4` and `Mirrored prod (GA4.0)` also be cleaned up somehow?
